src/ide/analysis_progress.rs

```diff
@@ -0,0 +1,180 @@
+    pub fn track_analysis<'a>(&self, beacons: impl Iterator<Item = &'a mut Beacon>) {
+        let gen_id = self.controller.next_generation_id();
+
+        self.controller.clear_active_snapshots();
+
+        beacons.enumerate().for_each(|(i, beacon)| {
+            self.controller.insert_active_snapshot(i);
+
+            let controller_ref: AnalysisProgressController = self.controller.clone();
+            beacon.on_signal(move || controller_ref.on_snapshot_deactivate(gen_id, i));
+        });
+
+        self.controller.start_analysis();
+    }
+}
+
+/// Controller used to send notifications to the client about analysis progress.
+/// Uses information provided from other controllers (diagnostics controller, procmacro controller)
+/// to assess if diagnostics are in fact calculated.
+#[derive(Debug, Clone)]
+pub struct AnalysisProgressController {
+    notifier: Notifier,
+    /// ID of the diagnostics "generation" - the scheduled diagnostics jobs set.
+    /// Used to filter out stale threads finishing when new ones (from newer "generation")
+    /// are already in progress and being tracked by the controller.
+    generation_id: Arc<AtomicU64>,
+    /// Sequential IDs of state snapshots from the current generation, used to track their status
+    /// (present meaning it's still being used)
+    active_snapshots: Arc<Mutex<HashSet<usize>>>,
+    id_generator: Arc<IdGenerator>,
+    /// If `true` - a request to procmacro server was submitted, meaning that analysis will extend
+    /// beyond the current generation of diagnostics.
+    did_submit_procmacro_request: Arc<AtomicBool>,
+    /// Indicates that a notification was sent and analysis (i.e. macro expansion) is taking place.
+    analysis_in_progress: Arc<AtomicBool>,
+    /// Loaded asynchronously from config - unset if config was not loaded yet.
+    /// Has to be set in order for analysis to finish.
+    procmacros_enabled: Arc<Mutex<Option<bool>>>,
+}
+
+impl AnalysisProgressController {
+    pub fn tracker(&self) -> AnalysisProgressTracker {
+        AnalysisProgressTracker { controller: self.clone() }
+    }
+
+    pub fn new(notifier: Notifier) -> Self {
+        let id_generator = Arc::new(IdGenerator::default());
+        Self {
+            notifier,
+            id_generator: id_generator.clone(),
+            active_snapshots: Arc::new(Mutex::new(HashSet::default())),
+            did_submit_procmacro_request: Arc::new(AtomicBool::new(false)),
+            analysis_in_progress: Arc::new(AtomicBool::new(false)),
+            procmacros_enabled: Arc::new(Mutex::new(None)),
+            generation_id: Arc::new(AtomicU64::new(id_generator.unique_id())),
+        }
+    }
+
+    pub fn set_did_submit_procmacro_request(&self, value: bool) {
+        self.did_submit_procmacro_request.store(value, Ordering::SeqCst);
+    }
+
+    /// Allows to set the procmacro configuration to whatever is in the config, upon loading it.
+    pub fn set_procmacros_enabled(&self, value: bool) {
```

Comment:
```suggestion
    pub fn on_config_change(&self, config: &Config) {
```

More future proof and it is clear where this info is coming from.